helm-chart/openj9-jitserver-chart/templates/deployment.yaml

```diff
@@ -0,0 +1,82 @@
+{{- /*******************************************************************************
+* Copyright (c) 2020, 2020 IBM Corp. and others
+*
+* This program and the accompanying materials are made available under
+* the terms of the Eclipse Public License 2.0 which accompanies this
+* distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+* or the Apache License, Version 2.0 which accompanies this distribution and
+* is available at https://www.apache.org/licenses/LICENSE-2.0.
+* 
+* This Source Code may also be made available under the following
+* Secondary Licenses when the conditions for such availability set
+* forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+* General Public License, version 2 with the GNU Classpath
+* Exception [1] and GNU General Public License, version 2 with the
+* OpenJDK Assembly Exception [2].
+* 
+* [1] https://www.gnu.org/software/classpath/license.html
+* [2] http://openjdk.java.net/legal/assembly-exception.html
+* 
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+*******************************************************************************/ -}}
+
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ include "openj9-jitserver-chart.fullname" . }}
+  labels:
+    {{- include "openj9-jitserver-chart.labels" . | nindent 4 }}
+spec:
+  replicas: {{ .Values.replicaCount }}
+  selector:
+    matchLabels:
+      {{- include "openj9-jitserver-chart.selectorLabels" . | nindent 6 }}
+  template:
+    metadata:
+      labels:
+        {{- include "openj9-jitserver-chart.labels" . | nindent 8 }}
+        {{- include "openj9-jitserver-chart.selectorLabels" . | nindent 8 }}
+    spec:
+    {{- with .Values.imagePullSecrets }}
+      imagePullSecrets:
+        {{- toYaml . | nindent 8 }}
+    {{- end }}
+      serviceAccountName: {{ include "openj9-jitserver-chart.serviceAccountName" . }}
+      securityContext:
+        {{- toYaml .Values.podSecurityContext | nindent 8 }}
+      containers:
+        - name: {{ .Chart.Name }}
+          command: {{ .Values.image.command }}
+          securityContext:
+            {{- toYaml .Values.securityContext | nindent 12 }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          env:
+            {{- toYaml .Values.env | nindent 12 }}
+          ports:
+            - containerPort: 38400
```

Comment:
This is part of the port forwarding in k8s that routes traffic from a user-defined port into container port 38400. Hope the below explanation makes sense. 

The first step in `values.yaml`, users can change this port as desired. Link to the [line](https://github.com/eclipse/openj9-utils/pull/8/files/2d3fd90d1e096b4d7d3c6a0ca93c5b9d1a754b22#diff-fd55dbffa8fd3838e03e1ef277504f57dc9389c69c7a6a1bcd6d577a4e08c53cR62). 
``` yaml
# Values.yaml
service:
  type: ClusterIP
  port: USER_DEFINED_PORT_NUM
```

Second, the `USER_DEFINED_PORT_NUM` is passed to `service.yaml`. Then k8s forward the `USER_DEFINED_PORT_NUM` to the `targetPort` (i.e. the default JITServer port 38400). Link to the [line](https://github.com/eclipse/openj9-utils/pull/8/files/2d3fd90d1e096b4d7d3c6a0ca93c5b9d1a754b22#diff-023338ece1cf12d8a931357446553de7f96aab39a991b9eefdf870d53678df8dR32).

``` yaml
# service.yaml
spec:
  ports:
    - port: {{ .Values.service.port }}  <--- USER_DEFINED_PORT_NUM
      targetPort: 38400
```

We need to use this number since JITServer only listens to this port in Adopt images. The kube object `server` is exposed as an endpoint and can be accessed by other services via `JITSERVER_SERVICE_NAME:USER_DEFINED_PORT_NUM`. 

Third, in `deployment.yaml`, we need to specify which port needs to be opened (`containerPort`). K8s routes traffic from service `targetPort` to deployment `containerPort`. Note the `containerPort` and `targetPort` needs to match or traffic cannot be passed into the container.  Link to the [line](https://github.com/eclipse/openj9-utils/pull/8/files/2d3fd90d1e096b4d7d3c6a0ca93c5b9d1a754b22#diff-032786001c06ec1eb608174eb023a467eb51cafa70e9fff6fa60128c3bc0207cR57).

```yaml
# deployment.yaml
spec:
  template:
    spec:
      containers:
        - name: {{ .Chart.Name }}
          ports:
            - containerPort: 38400
              protocol: TCP
```

Reply:
> We need to use this number since JITServer only listens to this port in Adopt images.

Isn't this configurable at the JITServer using the `-XX:JITServerPort=` option (or whatever the exact spelling of that option is) ?

Reply:
Yes, port number can be changed at the server with `-XX:JITServerPort=<num>`

Reply:
The question comes to shall we allow users to choose the port number only at the top level of port forwarding, and we can handle the rest (forwarding from `-XX:JITServerPort` to the outside of the container)? 
Both achieve the same effect, but `-XX:JITServerPort` needs to be changed once `containerPort` is changed to other values. I think it is easier for users to only change the top level port number and use that port as the "JITServer port". It might increase complexity and possible error for users to handle port forwarding at every level.

Reply:
I agree with Cris' explanation. Changing the port number at container level by providing `-XX:JITServerPort=...` at jitserver level is not necessary. Users can use different ports at the service level should they want to.